tests/baselines/reference/inferenceOptionalPropertiesToIndexSignatures.js

```diff
@@ -0,0 +1,40 @@
+//// [inferenceOptionalPropertiesToIndexSignatures.ts]
+declare function foo<T>(obj: { [x: string]: T }): T;
+
+declare const x1: { a: string, b: number };
+declare const x2: { a: string, b: number | undefined };
+declare const x3: { a: string, b?: number };
+declare const x4: { a: string, b?: number | undefined };
+
+let a1 = foo(x1);  // string | number
+let a2 = foo(x2);  // string | number | undefined
+let a3 = foo(x3);  // string | number
+let a4 = foo(x4);  // string | number
```

Comment:
Hm - seems like this is the only questionable one.

Reply:
Effectively, because we don't have a separate `missing` type, `undefined` represents the missing state in optional properties, so we really should remove `undefined` when we're inferring to an index signature as index signatures represent the type of those properties that are present.

Reply:
Feels like things are getting a bit untenable without `missing`; any plausible implementation of `foo` with a legal `x4` could return `undefined`. We should revisit it.